В цьому домашньому завданні ми реалізуємо логістичну регресію на `numpy`.
Ці завдання допоможуть вам ґрунтовно засвоїти основні концепції логістичної регресії та реалізувати їх на практиці 🔥

#### Завдання 1: Реалізація функції сигмоїди
1. З використанням `numpy` напишіть функцію `sigmoid(z)` для обчислення значення сигмоїди згідно з формулою:
   $$
   \sigma(z) = \frac{1}{1 + e^{-z}}
   $$
2. Використовуючи цю функцію, обчисліть значення сигмоїди для наступних даних: $ z = [-2, -1, 0, 1, 2] $. Виведіть результат обчислень.


In [1]:
import numpy as np

In [3]:
z = np.array([-2, -1, 0, 1, 2])
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

sigmoid_values = sigmoid(z)
print("Sigmoid values:", sigmoid_values)

Sigmoid values: [0.11920292 0.26894142 0.5        0.73105858 0.88079708]


Значення сигмоїд лежать у межах від 0 до 1, з чого можна зробити висновок що функція реалізована правильно



#### Завдання 2: Реалізація функції гіпотези для логістичної регресії
1. Напишіть функцію `hypothesis(theta, X)`, яка обчислює гіпотезу для логістичної регресії, використовуючи функцію сигмоїди. Формула гіпотези:
   $$
   h_\theta(x) = \sigma(\theta^T x) = \frac{1}{1 + e^{-\theta^T x}}
   $$
2. Використайте функцію `hypothesis` для обчислення значень гіпотези для наступних даних:
   
   $\theta = [0.5, -0.5]$
   
   $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  Виведіть результат обчислень.


In [7]:
theta = np.array([0.5, -0.5])
X = np.array([[1, 2], [1, -1], [1, 0], [1, 1]])

def hypothesis(X, theta):
    return sigmoid(np.dot(X, theta))

hypothesis_values = hypothesis(X, theta)
print("Hypothesis values:", hypothesis_values)

Hypothesis values: [0.37754067 0.73105858 0.62245933 0.5       ]


Значення параметрів тета перемножуємо відповідно до кожного значення х і отримуємо лінійну комбінацію ознак і функція сигмоїди використовується як функція гіпотези

#### Завдання 3: Реалізація функції для підрахунку градієнтів фукнції втрат
1. Напишіть функцію `compute_gradient(theta, X, y)`, яка обчислює градієнти функції втрат для логістичної регресії. Формула для обчислення градієнта:
   $$
   \frac{\partial L(\theta)}{\partial \theta_j} = \frac{1}{m} \sum_{i=1}^{m} \left[ (h_\theta(x^{(i)}) - y^{(i)}) x_j^{(i)} \right]
   $$
2. Використайте функцію `compute_gradient` для обчислення градієнтів для наступних даних:

  $\theta = [0.5, -0.5]$

  $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  $y = [1, 0, 1, 0]$

  Виведіть результат обчислень.

In [33]:
theta = np.array([0.5, -0.5])
X = np.array([[1, 2], [1, -1], [1, 0], [1, 1]])
y = np.array([1, 0, 1, 0])

def compute_gradient(theta, X, y):
    m = len(y)
    predictions = hypothesis(X, theta)
    gradient = np.dot(X.T, predictions - y) / m
    return gradient

gradient = compute_gradient(theta, X, y)
print("Gradient:", gradient)

Gradient: [ 0.05776464 -0.36899431]


Ми отримали градієнт, який показує, як змінювати параметри тета для зменшення функції втрат.
Тобто щоб зменшити втрати:
- параметр тета з індексом 0 потрібно зменшити, бо градієнт додатній,
- параметр тета з індексом 1 потрібно збільшити, бо градієнт від’ємний.


#### Завдання 4: Реалізація повного батч градієнтного спуску

**Задача:**
1. Напишіть функцію `full_batch_gradient_descent(X, y, lr=0.1, epochs=100)`, яка реалізує алгоритм Full градієнтного спуску для логістичної регресії. Використовуйте такі формули:
   - Гіпотеза: $ h_\theta(x) = \sigma(\theta^T x) $
   - Оновлення параметрів: $ \theta_j := \theta_j - \alpha \frac{\partial L(\theta)}{\partial \theta_j} $
2. Використайте функцію `full_batch_gradient_descent` для обчислення параметрів моделі на наступних даних:

  $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  $y = [1, 0, 1, 0]$

  Увага! Матриця $X$ вже має стовпець одиниць і передбачається, що це. - стовпець для intercept - параметра зсуву.

  Виведіть результат обчислень.


In [34]:
def full_batch_gradient_descent(X, y, theta, learning_rate=0.01, epochs=1000):
    for _ in range(epochs):
        gradient = compute_gradient(theta, X, y)
        theta -= learning_rate * gradient
    return theta

full_batch_gradient_descent_result = full_batch_gradient_descent(X, y, theta)
print("Theta after full batch gradient descent:", full_batch_gradient_descent_result)
compute_gradient_result = compute_gradient(theta, X, y)
print("Computed gradient:", compute_gradient_result)

Theta after full batch gradient descent: [-0.17526785  0.69545615]
Computed gradient: [ 0.03747162 -0.0313299 ]


Значення тета змінились отже навчання відбулось. Ми мали зменшити перший та збільшити другий, що відбулось і при цьому нові градієнти функції втрат покращились

#### Завдання 5. Обчислення точності моделі

1. Напишіть функцію `predict_proba(theta, X)`, яка використовує знайдені параметри $\theta$ для обчислення ймовірностей належності поточного прикладу з даних до класу $y=1$ на основі значень $\sigma(\theta^T x)$.

2. Напишіть функцію `predict(theta, X, threshold=0.5)`, яка обчислює клас з передбаченої імовірності належності екземпляра до класу 1 з порогом 0.5. Тобто якщо ймовірність менше 0.5, то передбачаємо клас 0, інакше клас 1.

3. Напишіть функцію `accuracy(y_true, y_pred)`, яка обчислює точність моделі, визначивши частку правильно передбачених класів.

  Формула метрики Accuracy:
  $$
  \text{Accuracy} = \frac{\sum_{i=1}^{m} I(\hat{{y}^{(i)}} = y^{(i)})}{m}
  $$

  де $\hat{{y}^{(i)}}$ - передбачене значення класу, $I$ - індикаторна функція (яка дорівнює 1, якщо умова виконується, і 0 - якщо ні), $m$ - кількість прикладів.

4. Обчисліть з використанням даних в завданні 4 $X$, $y$ та обчислених коефіцієнтах $\theta$ та виведіть на екран:
  - передбачені моделлю імовірності належності кожного з екземплярів в матриці `X` до класу 1
  - класи кожного екземпляра з матриці `X`
  - точність моделі.

In [36]:
def predict_proba(theta, X):
    return hypothesis(X, theta)

def predict(theta, X, threshold=0.5):
    probabilities = predict_proba(theta, X)
    return (probabilities >= threshold).astype(int)

def accuracy(y_true, y_pred):
    return np.mean(y_true == y_pred)

predict_prob_results = predict_proba(full_batch_gradient_descent_result, X)
print("Predicted probabilities:", predict_prob_results)
predict_results = predict(full_batch_gradient_descent_result, X)
print("Predicted results:", predict_results)
accuracy_result = accuracy(y, predict_results)
print("Accuracy:", accuracy_result)

Predicted probabilities: [0.77129615 0.29510367 0.45629486 0.6271918 ]
Predicted results: [1 0 0 1]
Accuracy: 0.5


Ми отримали масив імовірностей [0.77129615 0.29510367 0.45629486 0.6271918 ] що обчислює ймовірність належності кожного об"єкта до значення 1
Тобто передбачувані результати будуть виглядати так [1, 0, 0, 1] при тому що фактичні дані [1, 0, 1, 0]
Модель поки що передбачає з імовірністю 50% тобто вгадує лише половину значень.